### ISSUE TITLE
all: more thorough testing of GOAMD64=v1 instruction limits

### ISSUE THREAD

**Issue by rsc** [golang/go#63872]
#63871, which found a PSHUFB (v2+) not guarded by any instruction set checks in a crypto/tls dependency, shows that we are not testing GOAMD64=v1 terribly well. That is, it is possible to use newer instructions in Go packages accidentally, without breaking any tests. If we are going to provide GOAMD64=v1 it seems like we should have a test that it works as advertised. 

(To be clear, it is OK to emit these instructions when GOAMD64=v1, which is our default build mode, but only if they are guarded by an instruction set check before use. I don't see any such check in that case.)

I don't believe we have any v1 hardware to run on, and keeping access to that hardware would become progressively difficult in the future. I've heard a suggestion of using QEMU, but that's fraught with all sorts of problems. Instead I wonder if we should set up a stronger, builder version of cmd/compile/internal/amd64's TestGoAMD64v1. That test does the following:

1. Replace any non-v1 instruction with a breakpoint instruction, found by disassembling the binary.
2. Set GODEBUG environment variables to instruct code not to use post-v1 features.
3. Run itself (its own test binary).

The stronger version would set some environment variable that would cause the assembler to do two things:

1. Replace all non-v1 instructions with breakpoints.
2. Replace CPUID with a call to a simulation function that sets the registers to look like only v1 features are present.

This would apply during all of all.bash, not just during cmd/compile/internal/amd64.test. And it wouldn't depend on our getting the GODEBUG variables correct. We'd need a table of instruction->feature set for the assembler. We could set up the same for v2, v3.

Might be a good fixit week activity.

**Comment by randall77** [comment]
We already have a test like that, in `cmd/compile/internal/amd64/versions_test.go`. It builds a binary with `v1`, clobbers all the non-v1 instructions in the binary with breakpoints, and then runs it.

To catch the example you found, the test binary would just need to invoke some `crypto/tls` operations. Currently the test binary doesn't exercise much other than the `math` and `math/bits` packages (and the `test` package).

**Comment by rsc** [comment]
Right, this would be a more complete version of that test (I mentioned it above). Instead of having to add new things to the specific test over time, it would just test everything.

**Comment by 9072997** [comment]
I notice some of [the builders](https://farmer.golang.org/builders) appear to be maintained by community members. I would be interested in maintaining a v1 builder. I have a Sun workstation with an AMD Opteron 144 processor. It's not fast, but it's got 8gb of ram and I expect I could get a modern Linux kernel on it. AFAIK, it's the first generation of x86-64 processors (AMD K8), so if it runs on that it should run on anything.

More generally, it might be a good idea to try to have a builder for the baseline of each microarchitecture level, which [Wikipedia tells me](https://en.wikipedia.org/wiki/X86-64#Microarchitecture_levels) is:
- x86-64-v1: 2003 AMD K8 architecture
- x86-64-v2: 2008 Intel Nehalem architecture
- x86-64-v3: 2013 Intel Haswell architecture (specifically the ones with AVX2)
- x86-64-v4: 2017 Intel Skylake-X architecture

I can cover x86-64-v1 (AMD K8) and probably also x86-64-v2 (Nehalem). [EC2 C4 instances](https://aws.amazon.com/ec2/instance-types/) (which are still available) are guaranteed to run on Haswell E5-2666 v3. Skylake-X seems a little harder, since it appears to be desktop/workstation only and it's too new for me to have one lying around.
